### ISSUE TITLE
We should not be creating new zones all the time.

### ISSUE THREAD

**Issue by mhevery** [angular/zone.js#208]
I am looking at this code:

```
/**
 * requestAnimationFrame is typically recursively called from within the callback function
 * that it executes.  To handle this case, only fork a zone if this is executed
 * within the root zone.
 */
function patchRequestAnimationFrame(obj, fnNames) {
  fnNames.forEach(function (name) {
    var delegate = obj[name];
    if (delegate) {
      global.zone[name] = function (fn) {
        var callZone = global.zone.isRootZone() ? global.zone.fork() : global.zone;
        if (fn) {
          arguments[0] = function () {
            return callZone.run(fn, this, arguments);
          };
        }
        return delegate.apply(obj, arguments);
      };

      obj[name] = function () {
        return global.zone[name].apply(this, arguments);
      };
    }
  });
};
```

And it looks like we always create a new zone for each callback. (I think I remember asking for it.) But the comment makes me think that that is not right. If we tail recurse into the function we will create zones which are infinitely deep and hence we will leak memory, which is why the comment is there to prevent such a thing from happening.  This is an issue because:

1) it makes `raf` special
2) It does not actually fix the issue, because another common thing to do is to do `setTimetout(..., 0)` from within itself, hence same problem. 

So my current thinking is that it should only create a new zone if the developer explicitly calls `fork` (which is how it works is Dart). Which means that we need to implement long stack trace in some other way.

**Comment by btford** [comment]
How would we implement long stack traces more cheaply? We still have to record that information somewhere, so in the case of tail recursion, we still have the same problem, right?

**Comment by mhevery** [comment]
Yes long stack traces are a problem. But I think the problem is more complicated. Suppose you have tail recursion such as this:

```
doWork();

function doWork() {
  // do something
  requestAnimationFrame(doWork);
}
```

The above code is very common, and will cause memory leak and infinitely deep stack traces. There is special code in zone to prevent, that, but it is not a general solution, since same could be done with setTimeout, or any other callback.

So my current thinking is that we should support tail recursion by only creating zones when calling fork explicitly. This forks, but what about long stack traces? We can't have infinitely long stack traces so we need a way to manage this. My current thinking is that we can do this in two ways

1) we only save last x callbacks in the stack trace.
2) we detect tail recursion and stop adding data to the stack trace.

I favor option 2, but am having a hard time coming up with an ideal algorithm for this. ThoughtsL
